src/ir/stack-utils.h

```diff
@@ -158,6 +106,66 @@ struct StackSignature {
     return params == other.params && results == other.results &&
            unreachable == other.unreachable;
   }
+
+  // Whether a block whose contents have stack signature `a` could be typed with
+  // stack signature `b`, i.e. whether it could be used in a context that
+  // expects signature `b`. Formally:
+  //
+  // [t1*] -> [t2*] <: [s1* t1'*] -> [s2* t2'*] iff
+  //
+  //  - t1'_i <: t1_i
+  //  - t2_i <: t2'_i
+  //  - s1_i <: s2_i
```

Comment:
> * It'd be good to mention how the number of elements in the input and output figures in the subtype. `_i` seems to suggest that the numbers have to exactly match, but it doesn't seem to be the case, right?

Yes, the number of elements there need to match exactly. Any extra elements on the right hand side would be in the `s` part and the left side is not allowed to have extra elements. I will write that out explicitly.

> * Why is it ok to be `s1 <: s2`? If those types just sit on stack untouched, shouldn't they be the same instead?

For example, it is valid for an empty block to have type `[funcref] -> [anyref]` because nothing needs to happen to the `funcref` it takes as input for it to be interpreted as an `anyref` instead.

> * Does this rule about `s1` and `s2` only apply if `a` and `b` are not unreachable? Because if `a` or both are unreachable, those don't seem to matter.  For example...

Yes, that's right. If `a` is `unreachable`, the next rule applies and allows any `s1*` and `s2*` on the right hand side.